### DISCUSSION TITLE
Report error: "implementation of `std::marker::Send` is not general enough" in route handler

### DISCUSSION THREAD

**Discussion by ghost** [tokio-rs/axum#2816] [Q&A]
I previously used the actix-web framework, but now I've switched to using axum. I encountered the following error during compilation. What could be the reason for this? 

```
error: implementation of `std::marker::Send` is not general enough
  --> kernel/src/service/mod.rs:22:28
   |
22 |         .route("/graphql", post(kernel::create_query))
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `std::marker::Send` is not general enough
   |
   = note: `std::marker::Send` would have to be implemented for the type `&str`
   = note: ...but `std::marker::Send` is actually implemented for the type `&'0 str`, for some specific lifetime `'0`

error: implementation of `std::marker::Send` is not general enough
  --> kernel/src/service/mod.rs:22:28
   |
22 |         .route("/graphql", post(kernel::create_query))
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `std::marker::Send` is not general enough
   |
   = note: `std::marker::Send` would have to be implemented for the type `&TaskManger<Cuid, RegistryInstance<Cuid, (), TaskStorage>, TaskStorage, dispatcher::Dispatcher, ()>`
   = note: ...but `std::marker::Send` is actually implemented for the type `&'0 TaskManger<Cuid, RegistryInstance<Cuid, (), TaskStorage>, TaskStorage, dispatcher::Dispatcher, ()>`, for some specific lifetime `'0`
```

Here is my router
```rust
    let manager = Arc::new(TaskManager::default());
    let app = Router::new()
        .route("/graphql", post(kernel::create_query))
        .with_state(manager)
        .layer(ServiceBuilder::new().layer(TraceLayer::new_for_http()));
```

Here is my handler
```rust
pub async fn create_query(
    State(manager): State<Arc<TaskManager>>,
    body: String,
) -> impl IntoResponse {
    let query_id = match manager.create(&body).await {
        Ok(q) => q,
        Err(task_manager::Error::BadQuery(e)) => {
            tracing::info!("Request body({}) is invalid: {:?}", body, e);
            return Err((StatusCode::BAD_REQUEST, "Invalid request"));
        }
        _ => unreachable!(),
    };
    Ok(Json(json!({ "queryId": query_id.to_string() })))
}
```

Thanks

**Comment by mladedav** [comment]
Hi.

It seems your `TaskManger` isn't `Sync`. It's hard to tell without seeing the structure.

You might also want to try adding `#[debug_handler]` on your handler, it could provide a better error message.

**Reply by ghost** [reply]
I added the `#[debug_handler]`, but it didn't help. There is no new error message.

And according to the previous error message, it seems that it's a lifetime issue instead of `Sync`?

```
`std::marker::Send` would have to be implemented for the type `&TaskManger......
...but `std::marker::Send` is actually implemented for the type `&'0 TaskManger<....

```

**Reply by ghost** [reply]
And I tried to use `unsafe impl Send for` to impl `Send` for my `TaskManger` structure, it didn't help either.

**Reply by ghost** [reply]
Upon closer inspection, I discovered that invoking my handler directly within a `tokio::spawn` results in the same error, whereas encapsulating it within a `tokio::task::spawn_local` resolves the issue.

This discrepancy suggests a potential correlation with the `Send` trait ???

**Reply by mladedav** [reply]
`Sync` is implemented for `T` when `Send` is implemented for `&T`. I think that the error message is trying to say that `Send` is implemented for only some lifetimes of the reference, not all of them, e.g. only for `&'static T`. This would imply that `Sync` is not implemented.

Can you show what the Manager structure looks like? I think that would be the most helpful along with checking whether it implements `Send`/`Sync`.
